### ISSUE TITLE
Release: APT packaging

### ISSUE THREAD

**Issue by jmjwozniak** [swift-lang/swift-t#60]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Produce APT package for Swift/T

```

Original issue reported on code.google.com by `wozniak....@gmail.com` on 2 Nov 2012 at 6:51
* Blocked on: #571

_Copied from original issue: jmjwozniak/exm-issues#335_

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Yadu expressed interest
```

Original comment by `wozniak....@gmail.com` on 6 Nov 2012 at 3:45

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
I think issue 337 is a prerequisite for this, as we'll have to support Ubuntu's 
default TCL layout.
```

Original comment by `tim.g.ar...@gmail.com` on 6 Nov 2012 at 5:26

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_



Original comment by `wozniak....@gmail.com` on 25 Jan 2013 at 6:18
* Added labels: **Release-0.3.0**
* Removed labels: **Release-0.2.0**

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
I have c-utils packaged. Can someone please install and check ?

You could see where the files go with dpkg -L c-utils (after install *if* it 
works)

The next issue I see is the way adlb and everything else is configured upfront 
with c-utils location. Here we have the different files going to different 
places,
like lib/libexmcutils.so and headers(log.h, heap.h ...) to /include.



```

Original comment by `yadudoc1...@gmail.com` on 21 Feb 2013 at 2:25

Attachments:
* [c-utils_0.1.1-1_i386.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-4/c-utils_0.1.1-1_i386.deb)

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Awesome, nice work.  One day I'll have to figure out how to make an apt package 
myself.

I was able to get it to install on my system with no problems, although haven't 
tried using it yet.

It's installing files to /include and /lib.  I think this should by convention, 
instead be /usr/include and /usr/lib.  We may also want to have a subdirectory 
under /usr/include/exm-c-utils to avoid potential filename conflicts (table.h, 
list.h, etc are pretty generic names).  It seems like many libraries have their 
own subdirectories in include.

I googled and it appears that there's already an ubuntu package called cutils 
without the hyphen. Maybe we should call the package exm-c-utils to avoid 
confusion?
```

Original comment by `tim.g.ar...@gmail.com` on 21 Feb 2013 at 3:27

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Tim, I can share the scripts. The packaging is easy, but the guides are not.

Is this what we need :
/usr/lib:
libexmcutils.so

/usr/lib/c-utils/include:
c-utils-config.h  list2.h    list_lp.h  rbtree.h    tools.h
c-utils.h         list.h     list_sp.h  reader.h    tree-common.h
exm-string.h      list_i.h   log.h      table.h     tree.h
heap.h            list_ip.h  memory.h   table_ip.h  version.h
jenkins-hash.h    list_l.h   profile.h  table_lp.h


The renaming is okay, I can do that.
```

Original comment by `yadudoc1...@gmail.com` on 21 Feb 2013 at 4:42

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
That looks good to me, other packages seem to follow that layout.
```

Original comment by `tim.g.ar...@gmail.com` on 21 Feb 2013 at 5:26

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Wow!  This will really help get users running on their own.  
I will try it.

```

Original comment by `wozniak....@gmail.com` on 21 Feb 2013 at 5:57

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
You can put the scripts in the distro directory
```

Original comment by `wozniak....@gmail.com` on 21 Feb 2013 at 6:12

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
The *so files generated are missing symbols. There is a whole different set of 
steps for shared object libraries, which I missed. Please do not use the 
earlier deb.

Justin, thanks for noting that.
```

Original comment by `yadudoc1...@gmail.com` on 21 Feb 2013 at 6:48

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Justin, I am seeing an error from lintian which says : 
sharedobject-in-library-directory-missing-soname for lib/libexmcutils.so

"readelf -d libexmcutils.so | grep SONAME" returns nothing, for both the lib 
built by debuild as well as the one built by the exm-setup.zsh script. Is this 
an issue?
 
```

Original comment by `yadudoc1...@gmail.com` on 16 Mar 2013 at 12:14

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Does using ld's -soname=name fix this?
```

Original comment by `wozniak....@gmail.com` on 26 Mar 2013 at 4:59

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Turbine trunk is failing to build from the debuild system. I have it building 
normally using the setup, configure, make and I'm passing the same options to 
the debuild system.

I could really use some help here.

The logs and rules are attached. For now the c-utils and lb packaging is done.
Would be great if someone could try installing.




```

Original comment by `yadudoc1...@gmail.com` on 9 Apr 2013 at 8:35
* Changed state: **Accepted**

Attachments:
* [packager.log](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-13/packager.log)
* [turbine.rules](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-13/turbine.rules)
* [exm-c-utils_0.1.1-1_i386.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-13/exm-c-utils_0.1.1-1_i386.deb)
* [exm-lb_0.1.1-1_i386.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-13/exm-lb_0.1.1-1_i386.deb)

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_



Original comment by `wozniak....@gmail.com` on 24 May 2013 at 4:08
* Added labels: **Release-0.4.0**
* Removed labels: **Release-0.3.0**

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_



Original comment by `wozniak....@gmail.com` on 12 Jul 2013 at 9:54
* Added labels: **Release-0.5.0**
* Removed labels: **Release-0.4.0**

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_



Original comment by `wozniak....@gmail.com` on 1 Oct 2013 at 2:25
* Added labels: **Release-0.6.0**
* Removed labels: **Release-0.5.0**

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Turbine trunk packaging blocked due to Tcl8.6 not being found by configure.
(adding a link to issue)
```

Original comment by `yadudoc1...@gmail.com` on 12 Nov 2013 at 6:05
* Now blocked on: #571

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
I'm trying to place a symlink /usr/sbin/turbine pointing to the actual script 
/usr/sbin/exm/turbine/bin/turbine. When invoked this fails with the following 
error:

turbine -V
++ dirname /usr/sbin/turbine
+ TURBINE_BIN=/usr/sbin
+ [[ -z '' ]]
++ cd /usr/sbin/..
++ /bin/pwd
+ TURBINE_HOME=/usr
+ source /usr/scripts/turbine-config.sh
/usr/sbin/turbine: line 75: /usr/scripts/turbine-config.sh: No such file or 
directory
+ [[ 1 != 0 ]]
+ echo 'Turbine installation is broken!'
Turbine installation is broken!
+ exit 1

But if invoked with the full path /usr/sbin/exm/turbine/bin/turbine ,it works 
just fine. So the package installation was almost successful. Can this be 
corrected in the turbine script ?

I'm attaching the deb's.
```

Original comment by `yadudoc1...@gmail.com` on 12 Nov 2013 at 11:22

Attachments:
* [c-utils_0.6.0-1_amd64.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-18/c-utils_0.6.0-1_amd64.deb)
* [adlb_0.6.0-1_amd64.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-18/adlb_0.6.0-1_amd64.deb)
* [turbine_0.6.0-1_amd64.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-18/turbine_0.6.0-1_amd64.deb)

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Try using readlink and see if that helps
```

Original comment by `wozniak....@gmail.com` on 12 Nov 2013 at 11:32

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Can someone test these packages, please ?
4 packages attached. 
```

Original comment by `yadudoc1...@gmail.com` on 13 Nov 2013 at 5:16

Attachments:
* [c-utils_0.6.0-1_amd64.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-20/c-utils_0.6.0-1_amd64.deb)
* [adlb_0.6.0-1_amd64.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-20/adlb_0.6.0-1_amd64.deb)
* [turbine_0.6.0-1_amd64.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-20/turbine_0.6.0-1_amd64.deb)
* [stc_0.6.0_all.deb](https://storage.googleapis.com/google-code-attachments/exm-issues/issue-335/comment-20/stc_0.6.0_all.deb)

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
ADLB should depend on MPICH, not c-utils.
```

Original comment by `wozniak....@gmail.com` on 14 Nov 2013 at 3:31

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Turbine should depend on ZSH
STC should depend on Java

```

Original comment by `wozniak....@gmail.com` on 14 Nov 2013 at 3:41

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
With the deps, it basically works! 

The only remaining issue is to fix TCLLIBPATH

By default, I get: 

$ turbine -V hi.tcl
...
++ 
TCLLIBPATH+=/home/yadu/ubuntu/turbine-0.6.0/debian/turbine/usr/sbin/exm/turbine/
lib
...
+ mpiexec -n 3 /usr/bin/tclsh8.6 hi.tcl
can't find package turbine 0.3.0

The /home/yadu entry is wrong.  Note that this path is space-separated (because 
it is a Tcl list). 

I was able to fix this by setting:

$ export TURBINE_USER_LIB=/usr/sbin/exm/turbine/lib

```

Original comment by `wozniak....@gmail.com` on 14 Nov 2013 at 3:51

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Also, debugging should be off in our distributions.  Logging should be on. 
```

Original comment by `wozniak....@gmail.com` on 14 Nov 2013 at 3:52

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
I've fixed these : 
* ADLB now depends on mpich2, and not on c-utils
* Turbine depends on zsh
* stc depends on openjdk-7-jre
* TCLLIBPATH issue fixed with additional configure option

Isn't debugging off, and logging on, by default in the configure script? Where 
do I
where do I change this ?
 

```

Original comment by `yadudoc1...@gmail.com` on 14 Nov 2013 at 7:17

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
See #574
```

Original comment by `wozniak....@gmail.com` on 14 Nov 2013 at 7:34

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_

```
Links to packages :
http://swift.rcc.uchicago.edu:8042/deb_packages/adlb_0.6.0-1_amd64.deb
http://swift.rcc.uchicago.edu:8042/deb_packages/c-utils_0.6.0-1_amd64.deb
http://swift.rcc.uchicago.edu:8042/deb_packages/turbine_0.6.0-1_amd64.deb
http://swift.rcc.uchicago.edu:8042/deb_packages/stc_0.6.0_all.deb

I've rebuilt these with all debug options turned off for turbine, and no extra 
settings were passed to the configure scripts to turn off logging. Please try 
and 
let me know if these work for you.

I'm looking into hosting the deb packages on a ppa through launchpad.
```

Original comment by `yadudoc1...@gmail.com` on 14 Nov 2013 at 10:43

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_



Original comment by `wozniak....@gmail.com` on 6 Nov 2014 at 4:29
* Added labels: **Release-0.8.0**
* Removed labels: **Release-0.6.0**

**Comment by jmjwozniak** [comment]
_From @GoogleCodeExporter on April 22, 2015 18:55_



Original comment by `wozniak....@gmail.com` on 12 Mar 2015 at 2:30
* Added labels: **Release-0.9.0**
* Removed labels: **Release-0.8.0**

**Comment by jmjwozniak** [comment]
Basheer, go ahead and try the existing scripts.  Yadu, can you tell him how to run them?

**Comment by basheersubei** [comment]
I'm using `sudo dpkg -i package_name.deb` to install the deb packages.
* `c-utils` installed successfully (after installing some mpich2 dependencies).
* `stc` installed successfully
* `adlb` failed because [it's looking for a dependency that no longer exists](https://gist.github.com/basheersubei/3c64641da08027f8bd18) (at least I can't find it using apt-get).
* `turbine` depends on `adlb` so it can't install until `adlb` is installed.

**Comment by basheersubei** [comment]
it seems that [`libmpich2-3`](https://launchpad.net/ubuntu/trusty/i386/libmpich2-3/1.4.1-4.2ubuntu2) is obsolete now on Trusty, and I couldn't even install it manually because it gives a dependency error on `libcr0` (which I have installed...) I'll take a look inside the `adlb` deb file...

**Comment by jmjwozniak** [comment]
That should refer to whatever MPICH is current: http://lists.mpich.org/pipermail/discuss/2013-November/001752.html

**Comment by jmjwozniak** [comment]
I meant to tell you: you should test this on a VM to avoid clobbering your OS.

**Comment by basheersubei** [comment]
ok I just managed to get a nice VM setup and I removed the `libmpich2-3` dependency (did not replace it with anything) from the `control` file inside the `adlb` deb, and now installing it with `dpkg -i` is successful. I did the exact same for the `turbine` deb, with same result.

I'm attempting to actually run `stc` and `turbine` to see if it works, but there's some dependencies missing:

* `zsh`
* `java`
* `swig`

**Comment by basheersubei** [comment]
@jmjwozniak how do I figure out which `java` they depend on? Would the `default-jre` for Ubuntu work? Or do we need a jdk like `openjdk-6`?

**Comment by basheersubei** [comment]
nvm, I think `default-jre` installs `openjdk-7`. It worked (at least `stc` did).

**Comment by basheersubei** [comment]
@jmjwozniak I'm getting the [same error](https://gist.github.com/basheersubei/28a7bb6a0e81e1549d00) as you were on Tuesday: when running `turbine`, it says it can't find the `turbine` package.

I noticed the version of turbine I have is 0.4.0, but the `stc`-generated file lists `turbine` as 0.3.0. I tried changing the line `package require turbine 0.3.0` to `package require turbine 0.4.0`, but I got the same error.

**Comment by basheersubei** [comment]
I tried changing the `turbine-version.txt` file where `stc` was installed (i.e. `/usr/sbin/exm/stc`), and recompiled the .swift file, but same results...

**Comment by basheersubei** [comment]
also, a lot of these scripts inside the packages are riddled with absolute paths (which will not work on any other machines besides yadu's). Maybe a separate ticket for those?

**Comment by basheersubei** [comment]
upon inspection of the deb files, it seems all they do is copy the files over to `/usr/sbin/exm/` and check dependencies. This seems to be very different from the installation method described in the Swift/T guide or from the repo README. Why don't we make a meta-package deb (maybe call it swift-t or exm) that installs all these 4 deb packages (`c-utils`, `stc`, `adlb`, and `turbine`), creates a standard exm-settings file, then runs the install script or something? Am I thinking on the right track?

I think that would be the best way for users to install swift-t, as one single deb package.

**Comment by jmjwozniak** [comment]
I just fixed the version number issue, sorry about that.  You need to run 'ant install' again and it should work.

**Comment by jmjwozniak** [comment]
A meta package is fine.  We do want to be able to install only a subset of swift-t, for cases where you just want the library or runtime.  The idea of de-duplicating installation code is great if you can do it.  I have not looked at APT in detail so feel free to propose something in a branch or fork and Yadu or I will try it.

**Comment by jmjwozniak** [comment]
BTW, if possible, let's call the ADLB package 'adlb-x' or something- we should reserve 'adlb' for the original ADLB inventor (Lusk)!  Don't worry about renaming libadlb.so or anything (for now).

**Comment by basheersubei** [comment]
@jmjwozniak I'm confused about what you fixed (fd6ce496a2cf2d685ec06d343f01ee86a5b31c90). The deb package has that file in a different directory (`/usr/sbin/exm/stc/turbine-version.txt`), and I'm not building or installing anything manually (I'm just running `dpkg -i` on these deb files). Where do I run `ant install`?

these deb packages have turbine binaries version 0.4.0. I tried changing everywhere where turbine-version is set to 0.4.0, but it still doesn't work, same error.

**Comment by basheersubei** [comment]
I also tried setting the value for version to 0.6.0 everywhere but it didn't work. Should I start over using the most recent version of the libraries (from the git repo)?

**Comment by basheersubei** [comment]
Just to be clear, when I build everything (using the scripts) from the recent version (HEAD of this repo), everything works fine (both STC and Turbine). I'm not sure what's up with the binaries in the deb packages, so I think I'll just create a deb package using binaries from the newest version.

**Comment by jmjwozniak** [comment]
The version number is written in the TIC by STC.  The TIC has a Tcl package require statement for Turbine that must be satisfied.  You may want to experiment with a toy Tcl package to see how this works (and fails).

ant install is used to compile and install STC from the Java source.  The deb scripts could have some bug regarding this feature, which is intended to prevent users from trying to run TIC from a newer STC on and older Turbine.

Definitely use the newest version of the source code.

**Comment by basheersubei** [comment]
@jmjwozniak just to double-check: the build scripts in the `dev` folder compile all four packages for `amd64` (x86_64) architecture, right?

**Comment by basheersubei** [comment]
I'm naming the packages as such:
* exm-c-utils
* exm-stc
* exm-turbine
* exm-adlb-x

Let me know if I should do something else for the naming

**Comment by basheersubei** [comment]
also, i'm attempting to use [pbuilder](https://wiki.ubuntu.com/PbuilderHowto) to set-up chroot clean environments to test the debian packages

**Comment by jmjwozniak** [comment]
exm-c-utils is good; for the others I would like to keep it simple- adlb-x, turbine, stc.  I would like to put these in the official repositories if possible so let me know if there are conflicts.

**Comment by basheersubei** [comment]
sounds good (I can't see any name conflicts on the official `apt-get` repositories). What should I call the meta-package?

**Comment by jmjwozniak** [comment]
swift-t

**Comment by basheersubei** [comment]
i'm a little confused by the version numbers. Are all the four packages the same version? And is it `0.8.0`?

**Comment by jmjwozniak** [comment]
They are different- see here: https://sites.google.com/site/exmproject/development/swift-t---releases

**Comment by basheersubei** [comment]
it says I need to request access (which I just did), but the email failed to be sent.

**Comment by basheersubei** [comment]
ok, there seems to be a problem with the way the packages are structured. (I could be wrong here but this is what I think the situation is):

the usual way programs are installed is to place the binaries in `/usr/bin`. Any other conf stuff they need is placed in locations like `/usr/conf` or `/usr/etc` and such. So you end up with a directory tree looking like this:

    root
       |
       | -- /usr
              |
              | -- /bin  (has all the binary executables)
              |
              | -- /lib  (shared libraries .so and .a)
              | -- /conf  (any config scripts)
              | -- /include (.h header files)
              | -- /etc  (other files)


The way the packages are structured now (and I'll use `stc` as an example so we can be specific) creates some issues.

Essentially, there's some hardcoded inter-dependencies between the packages. `stc` expects the `turbine` installation to be in a specific place (which can be fixed by changing the path in `stc-env.sh`), but the installation directory structure is hardcoded. It expects there to be a folder called `turbine` with a `/bin` and `/etc` inside of it (which doesn't work with the structure I described above).

I see two fixes:
1- Just install everything in one directory like all 4 packages expect. Something like `/home/USERNAME/swift-t/` and then place each of the 4 packages as its own subdirectory.
2- Change the directory structure of all 4 packages to match the one I described above. This involves going through source code to change where stuff points to (files and directories etc.)

Let me know what you think (I'm not sure I'm on the right track). I don't think option 2 is a good idea, as I'm not even sure this is the installation structure we want. I'll go with option 1 for now, and see if I can get something working first (regardless of how neat the directories look).

**Comment by basheersubei** [comment]
For reference, [here's](https://github.com/basheersubei/exm-debians/tree/master/original-installation) what the original installation directory structure looks like, and [here's](https://github.com/basheersubei/exm-debians/tree/master/modified-structure) what it's "supposed" to look like ideally according to the above description.

**Comment by basheersubei** [comment]
ok, I've made basic packages that work (similar to the ones Yadu made, except with the newer binaries). The only fix I have to add is to write a postinstall script that adds the binary paths to the `PATH` environment variable (should be easy).

**Comment by basheersubei** [comment]
I tested this on a fresh install of Ubuntu Server 14.04, and it works fine with basic hello world swift script.

**Comment by basheersubei** [comment]
I can confirm it also works on an Amazon EC2 instance. Remaining possible issues:

- [ ] `lintian` gives a lot of warnings (mostly wrong permissions).
- [ ] adding the debs to some sort of repository (either a PPA or the official repository [in which case the lintian issue needs to be fixed first])
- [ ] the /bin directory is appended to the $PATH variable in `/etc/profile`, but it still requires the user to source that file (only once and for the current shell session they just installed the package from).

**Comment by jmjwozniak** [comment]
Sounds great!  Are any of the lintian issues blockers? 
What do I need to do to try it? 
Let's not worry too much about /bin - we can instruct users to add that to PATH manually.

**Comment by basheersubei** [comment]
sorry I thought I linked this before. Everything is in [this repo](https://github.com/basheersubei/exm-debians). The instructions in the README should cover (hopefully) all issues.

I will double-check the lintian issues if they're really important (I'll fix them soon)

**Comment by basheersubei** [comment]
## Serious Issues:
- [ ] [binary-or-shlib-defines-rpath](https://lintian.debian.org/tags/binary-or-shlib-defines-rpath.html): the binaries define their own RPATH, which, according to [this](https://wiki.debian.org/RpathIssue) is frowned upon (for most cases). In our case, it links to non-standard library paths, which is ok (?). Perhaps it's better if I were to put the `.so` files in `/usr/lib/swift-t/` as opposed to `/usr/lib/` (right now).
- [ ] [sharedobject-in-library-directory-missing-soname](https://lintian.debian.org/tags/sharedobject-in-library-directory-missing-soname.html): The `.so` files are missing sonames, which *might(?)* be ok because the `.so` files are not versioned and are only intended to be used within the same package (swift-t stuff). The above issue might fix this.

## Minor Issues:
- [ ] [jar-not-in-usr-share](https://lintian.debian.org/tags/jar-not-in-usr-share.html): are these hard-coded (if not, it should be simple to just move them to `/usr/share`).
- [ ] a bunch of bad permissions or wrong user owner. Pretty simple to fix.

**Comment by jmjwozniak** [comment]
Debians are done.
